drivers/scd30/scd30.c

```diff
@@ -0,0 +1,352 @@
+        scd30_get_param(dev, SCD30_STATUS, &state);
+    }
+
+    if(state != 1) {
+        return SCD30_NO_NEW_DATA;
+    }
+    ret = _scd30_read_data(dev, result);
+    if (ret != SCD30_OK) {
+        DEBUG("[scd30] Error reading values");
+        return ret;
+    }
+    return SCD30_OK;
+}
+
+int scd30_start_periodic_measurement(scd30_t *dev, uint16_t *interval,
+                                     uint16_t *apc)
+{
+    int ret = 0;
+
+    /* Check if input is valid */
+    if (!_is_valid_interval(interval)) {
+        DEBUG("[scd30] Interval value out of range\n");
+        return SCD30_INVALID_VALUE;
+    }
+    if (!_is_valid_press_comp(apc)) {
+        DEBUG("[scd30] Ambient pressure compensation value out of range\n");
+        return SCD30_INVALID_VALUE;
+    }
+
+    ret = scd30_set_param(dev, SCD30_INTERVAL, *interval);
+    if (ret != 0) {
+        return ret;
+    }
+    ret = scd30_set_param(dev, SCD30_START, *apc);
+    if (ret != 0) {
+        return ret;
+    }
+    return SCD30_OK;
+}
+
+int8_t scd30_stop_measurements(scd30_t *dev)
+{
+    uint8_t buffer[2];
+    uint16_t cmd = SCD30_STOP;
+    int ret;
+
+    buffer[0] = (uint8_t)(cmd >> 8);
+    buffer[1] = (uint8_t)cmd;
+    i2c_acquire(SCD30_I2C);
+    ret = i2c_write_bytes(SCD30_I2C, SCD30_I2C_ADDRESS, buffer, 2, 0);
+    i2c_release(SCD30_I2C);
+    if (ret != 0) {
+        DEBUG("[scd30]: Error stopping measurements.\n");
+    }
+    return SCD30_OK;
+}
+
+/**
+ *      intern function to read data
+ *
+ *      @param  dev      device
+ *		@param  result	 struct in which data read is stored
+ *      @return          SCD30_OK on success
+ */
+static int8_t _scd30_read_data(scd30_t *dev, scd30_measurement_t *result)
+{
+    uint8_t buffer[18];
+    int ret = 0;
+    uint16_t cmd = SCD30_DATA;
+    uint32_t temp;
+
+    buffer[0] = (uint8_t)(cmd >> 8);
+    buffer[1] = (uint8_t)cmd;
+    i2c_acquire(SCD30_I2C);
+    ret = i2c_write_bytes(SCD30_I2C, SCD30_I2C_ADDRESS, buffer, 2, 0);
+    if (ret != 0) {
+        i2c_release(SCD30_I2C);
+        return ret;
+    }
+    ret = i2c_read_bytes(SCD30_I2C, SCD30_I2C_ADDRESS, buffer, 18, 0);
+    if (ret != 0) {
+        DEBUG("[scd30] read_data: ret %i.\n", ret);
+        i2c_release(SCD30_I2C);
+        return ret;
+    }
+    i2c_release(SCD30_I2C);
+
+    if (_scd30_crc_check(buffer, 2) != 0) {
+        return SCD30_CRC_ERROR;
+    }
+    if (_scd30_crc_check(buffer + 3, 2) != 0) {
+        return SCD30_CRC_ERROR;
+    }
+    if (_scd30_crc_check(buffer + 6, 2) != 0) {
+        return SCD30_CRC_ERROR;
+    }
+    if (_scd30_crc_check(buffer + 9, 2) != 0) {
+        return SCD30_CRC_ERROR;
+    }
+    if (_scd30_crc_check(buffer + 12, 2) != 0) {
+        return SCD30_CRC_ERROR;
+    }
+    if (_scd30_crc_check(buffer + 15, 2) != 0) {
+        return SCD30_CRC_ERROR;
+    }
+    temp = byteorder_bebuftohl(&buffer[0]);
+    memcpy(&result->co2_concentration, &temp, 4);
```

Comment:
I'd avoid the memcpy, just use your old helper function

```C
static inline float _reinterpret_float(uint32_t raw_val)
{
    union {
        float float_val;
        uint32_t int_val;
    } to_float = {.int_val = raw_val};

    return to_float.float_val;
}
```

Reply:
Could you tell me why you recommend not to use memcpy ? (just curious)

Reply:
I thought we should avoid the library call, since this we basically just have to move one word of memory around, that can be a single instruction.

But turns out the compiler is [clever enough](https://godbolt.org/z/zzGEzn) to entirely remove the call to `memcpy` and will in fact generate the same code either way.

Reply:
Woah, that's pretty cool.